### PR TITLE
[JENKINS-5597] add option to follow symlinks during artifact archiving

### PR DESCRIPTION
**Pull request by mwinter69**
add an option to ArtifactArchiver that allows to configure follow symlink behaviour of the DirectoryScanner used by ArtifactArchiver. For newly created steps the default is now to not follow symlinks. Existing steps will work as before (also for pipeline).

See [JENKINS-5597](https://issues.jenkins-ci.org/browse/JENKINS-5597).

Minor improvement.

### Proposed changelog entries

* Entry 1: Add option to configure follow symlinks when archiving artifacts

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

**Review by Wadeck — COMMENTED** [review]
I am not sure we really want to have such feature, especially for security reasons. In the recent patch around illegal symlink (trying to escape the workspace) there was a discussion around removing completely the symlink feature to avoid such troubles.

@jglick any thoughts here?

**Comment by mwinter69** [comment]
actually it is now possible to disable the follow symlink. Before it was always enabled, causing some issues.

**Comment by daniel-beck** [comment]
If there is a security issue here, then it already exists today: _Invalid_ symlinks are archived as symlinks currently, so this should be no different from declaring all symlinks to be invalid today. (In fact, that was [SECURITY-162](https://jenkins.io/security/advisory/2015-02-27/#security-162-directory-traversal-from-artifacts-via-symlink)).

File: core/src/main/resources/hudson/tasks/ArtifactArchiver/config.jelly
Comment (LakshanSS) [inline core/src/main/resources/hudson/tasks/ArtifactArchiver/config.jelly]: Have to add a newline in the end of file

File: core/src/main/resources/hudson/tasks/ArtifactArchiver/config.properties
Comment (LakshanSS) [inline core/src/main/resources/hudson/tasks/ArtifactArchiver/config.properties]: Can add a new line in the end of file

File: core/src/main/resources/hudson/tasks/ArtifactArchiver/config_de.properties
Comment (LakshanSS) [inline core/src/main/resources/hudson/tasks/ArtifactArchiver/config_de.properties]: Add a new line in the end of the file

File: core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks.html
Comment (LakshanSS) [inline core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks.html]: add a new line in the end of the file

File: core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks_de.html
Comment (LakshanSS) [inline core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks_de.html]: add a new line in the end of the file

**Comment by jglick** [comment]
Indeed going forward I recommend skipping all symlinks (either reporting an error or logging a warning). Any project which actually needs to record a directory tree containing symlinks or other filesystem exotica is better advised to create a tarball in the workspace as part of the build script.

**Comment by AbhishekSharma20** [comment]
Hi @jglick @Wadeck ,
Can you please integrate these changes? It would be very helpful. Our work area uses recursive symlinks as a feature and cannot be removed since there are a lot of dependencies. There're many people affected by this issue, we have to put the manual effort in building custom jenkins.

If there is an enhancement at a later stage regarding removing symlinks traversal all together, then just changing the default value of the boolean variable controlling this behavior will serve the purpose.

-Abhishek

**Comment by jglick** [comment]
> we have to put the manual effort in building custom jenkins.

Seems unlikely; again, you can already just create a tarball during your build and archive that instead.

To make this more attractive and guide users to do the right thing, perhaps we could just offer an option to `ArtifactArchiver`. Something like:

```groovy
sh 'make build'
dir('build') {
  archiveArtifacts includes: '**/', archive: 'binaries.tgz'
}
```

which would do the tarring on the agent side, saving you the trouble of adding a `tar` or `zip` command to your build script. (The compression method could be inferred by the filename: `.zip`, `.tar`, `.tar.gz` / `.tgz`.) We could then consider making it an error if a symlink is encountered _without_ this option being set, with a system property to keep the problematic original behavior for a while before really deleting that ability.

**Comment by AbhishekSharma20** [comment]
Hi @jglick,
My issue was that my directory is structured in a very complex way containing many levels of symlinks (because of business requirements). archiving this way won't work as the workspace tree will still be created inside Jenkins. and it is that object (the Hashmap containing directoryStructure) which is getting infinite bloating up the memory. As of now, we have faked the work area to Jenkins to an empty directory inside our actual work area. Since this directory will not contain anything but artifacts to be archived, this issue will not come.

**Review by MRamonLeon — APPROVED** [review]
I'm fine with this change, the discussion about the use or not of symlinks is another topic I believe. So approving.

It would be great if you add the EOL suggested.

**Comment by fcojfernandez** [comment]
After reading @jglick , @Wadeck and @daniel-beck 's comment I have actual doubts if this PR should be merged or not. Maybe it's worth to open a discussion in the mailing list?

**Comment by alecharp** [comment]
I personally don't feel comfortable with this feature, I can only see security risks. I would prefer not to have this in Jenkins, symlinks in general don't feel like the correct solution for a build.

**Comment by mwinter69** [comment]
We have one case where a node build is creating many softlinks to other modules (even with cycles). Using find to get the the real number of files and folder returns like 63000 lines.
But due to the softlinking this is blown up to almost 1.3 million. In combination with wildcard patterns we see OOM exceptions in the ant directory scanner as also reported by @AbhishekSharma20 . With this feature there is no issue.

I agree that symlinks are dangerous when they lead to files outside of the workspace. It should not be possible to archive anything that is outside of the workspace (though it will not stop anyone to just copy the file instead of creating a softlink and archive the file this way).
But symlinks that stay within the workspace are totally ok I think and should not be forbidden. 

The solution by @jglick with an automatic archiving will not change anything to the OOM if the directory scanner is enabled to follow symlinks.

My PR will not make it any worse to the current situation but solves a problem when there are symlinks in the workspace.

**Comment by Wadeck** [comment]
I will try to spend some time to re-review this PR as it seems promising :)

**Comment by jglick** [comment]
> the workspace tree will still be created inside Jenkins. and it is that object (the Hashmap containing directoryStructure) which is getting infinite bloating up the memory

This just sounds like a bug. Are there steps to reproduce?

> As of now, we have faked the work area to Jenkins to an empty directory inside our actual work area. Since this directory will not contain anything but artifacts to be archived, this issue will not come.

That seems like a good workaround.

**Comment by jeffret-b** [comment]
I plan on reviewing in the next couple of days.

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (varyvol) [inline core/src/main/java/hudson/tasks/ArtifactArchiver.java]:
```suggestion
     * Indicate whether symbolic links should be followed or not
```

**Review by varyvol — APPROVED** [review]
Minor suggestion for the Javadoc, but the change looks good give it actually allows to disable following symlinks, which could make the situation better.

**Comment by jeffret-b** [comment]
I don't see that this PR causes any additional problems that don't already exist. It provides some additional capability which can be useful. I'm in favor of accepting this PR (with a little bit of change).

Ultimately, I agree with Jesse and others that handling symlinks is just a bad idea. Tar is a much better solution for archiving symlinks. Zip doesn't really know what to do with symlinks, especially on Windows.

But, a lot of people use the Archive Artifact capability already and this would be a step in the right direction to better use. I would prefer to make not following symlinks the default, but that is probably too much of a change in behavior. Maybe we could change the default later on.

I don't see that we need to hold up this PR for some better, more comprehensive solution that doesn't yet exist.

**Review by jeffret-b — CHANGES_REQUESTED** [review]
This behavior really deserves an automated test. It looks like there are existing tests in ArtifactArchiverTest that should be easily modifiable to test this additional behavior. I can help suggest what to do if needed.

I favor approving this PR and merging it in, but I would like to have a test for it.

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (jeffret-b) [inline core/src/main/java/hudson/tasks/ArtifactArchiver.java]: @daniel-beck, if this is false and with the readResolve() setting to true on null, does that retain old behavior for existing projects and default to not following symlinks for new ones?

We should definitely have that behavior and then I'm even more in strongly in favor of this PR.

Comment (daniel-beck) [reply core/src/main/java/hudson/tasks/ArtifactArchiver.java]: @jeffret-b Mostly.

This being `Boolean` and `readResolve` combine to set the the default when loading existing projects. This is definitely the correct approach to not break everyone's existing projects.

This sets the programmatic default for new instances, but the UI needs to be considered as well. https://github.com/jenkinsci/jenkins/pull/3947/files#diff-16a525c7f3ec8f3f359bc969708f920cR50 would make a different value here next to useless (modulo Pipeline snippet generator and such).

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (daniel-beck) [inline core/src/main/java/hudson/tasks/ArtifactArchiver.java]:
```suggestion
        }
```

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (daniel-beck) [inline core/src/main/java/hudson/tasks/ArtifactArchiver.java]: How confident are we that this never breaks thanks to `readResolve`?

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (daniel-beck) [inline core/src/main/java/hudson/tasks/ArtifactArchiver.java]:
```suggestion
        this.followSymlinks = followSymlinks;
```

**Comment by daniel-beck** [comment]
Overall I like this a lot. I would even argue (as does @jeffret-b ) that not following symlinks is a superior approach to doing so.

Given how central archiving of artifacts is to Jenkins, this deserves autotests. Otherwise I see nothing wrong with this.

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (jeffret-b) [reply core/src/main/java/hudson/tasks/ArtifactArchiver.java]: So, change that line to `false` also, and it should work, right? That's good.

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (jeffret-b) [reply core/src/main/java/hudson/tasks/ArtifactArchiver.java]: I'm pretty confident. If we had another test or two, I'd be sufficiently confident.

**Comment by mwinter69** [comment]
added some tests.
I included and update of ant to 1.10.7 as since 1.10.2 the ant directory scanner is able to properly detect symlinks on windows.
Haven't yet tested behaviour of pipeline jobs

**Comment by jeffret-b** [comment]
I'm glad to see the addition of tests. I plan on reviewing this again tomorrow.

**Comment by daniel-beck** [comment]
@jglick Is there guidance on how Pipeline steps should evolve to add options with different default behavior?

For traditional job types, we can `readResolve` into legacy behavior, but if we change the new field `followSymlinks` to `false`, IIUC, the `archiveArtifacts` step will change behavior.

Having a different default behavior for new freestyle jobs vs. new pipelines (all pipelines really) is also weird.

**Comment by jglick** [comment]
> Is there guidance on how Pipeline steps should evolve to add options with different default behavior?

The [usual guide](https://jenkins.io/doc/developer/plugin-development/pipeline-integration/#handling-default-values).

I have not followed the discussion here about default values. Whatever the constructor picks, that will be the default value for Pipeline. I would certainly not advise changing the default behavior, as you could break existing scripts. I have no opinion about the behavior for non-Pipeline projects.

File: core/src/main/java/hudson/tasks/ArtifactArchiver.java
Comment (jeffret-b) [reply core/src/main/java/hudson/tasks/ArtifactArchiver.java]: If I understand the flow of discussion correctly, to maintain backwards compatibility for pipeline, we need to keep this as `true` and keep the new UI setting in config.jelly defaulting to false.

File: core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks.html
Comment (jeffret-b) [inline core/src/main/resources/hudson/tasks/ArtifactArchiver/help-followSymlinks.html]: I recommend simplifying this. I don't think the user really needs to know about Ant and Directory Scanner. I would probably remove the help text entirely, as I think "Follow symbolic links" is sufficiently explanatory.

**Review by jeffret-b — APPROVED** [review]
I'm glad this now has the two new tests. That helps a lot.

As I'm understanding it, this now works correctly to maintain compatibility with existing jobs and with pipelines but defaults in the UI to not follow symlinks. That's probably as good as we can get for reducing symlink usage at this point.

I left a comment suggesting simplifying or removing the help text.

I support this change. It will move us towards reducing symlink usage, which is a good thing.

**Comment by daniel-beck** [comment]
> keep the new UI setting in config.jelly defaulting to false.

That's the question – it looks like we cannot just change the programmatic default (thanks pipeline!); I wonder what's less bad:

* Having the inferior behavior be the default for all job types (terrible but consistent)
* Having the inferior behavior only be the default for pipelines (terribly inconsistent)

The former is probably better. Even if it still sucks.

**Comment by jeffret-b** [comment]
>     * Having the inferior behavior be the default for all job types (terrible but consistent)
> 
>     * Having the inferior behavior only be the default for pipelines (terribly inconsistent)
> 
> 
> The former is probably better. Even if it still sucks.

+1 for the second choice

**Comment by jglick** [comment]
> it looks like we cannot just change the programmatic default (thanks pipeline!)

Note that this would not be 100% safe even for non-Pipeline projects: `readResolve` system only works insofar as you are deserializing an existing `config.xml`. It would not work for people defining projects using (say) `job-dsl`, which while not the common case is not all that exotic either.

**Comment by jeffret-b** [comment]
So, what's the necessary level of backwards of compatibility? Can we set the UI to default to not following symlinks and have the constructor default to following? Will that be sufficient?

I'd like to keep this moving forward as best we can to get us into a better position.

**Comment by mwinter69** [comment]
> 
> So, what's the necessary level of backwards of compatibility? Can we set the UI to default to not following symlinks and have the constructor default to following? Will that be sufficient?
> 
> I'd like to keep this moving forward as best we can to get us into a better position.

This is how it is at the moment. The UI defaults to false for freestyle and pipeline (snippet generator). As the default is true in the class and readresolve also sets it to true this will ensure full backward compatibility for existing jobs and pipelines I think.

**Comment by jeffret-b** [comment]
@Wadeck , you marked this one as needing more reviews and indicated you would review. Would you please add yours when you can?

**Comment by rsandell** [comment]
It is a bit unclear to me what the next steps are and if there is any consensus.

**Comment by jeffret-b** [comment]
I think it's ready to go. @Wadeck  said it was going to review it and that it needed more reviews so I'm trying to give him the chance. Otherwise, it has three reviews and I don't think it has any outstanding concerns.

**Comment by Wadeck** [comment]
I am perfectly fine with one person from our team to have approved it :) (I have change my initial comment to highlight that fact)

**Comment by daniel-beck** [comment]
Seems fine. A bit annoying that Pipeline / Job DSL prevent us from really changing the global default to something sane.

Right now,

1. Pipeline and Job DSL will by default _follow_
2. Existing build steps will continue to _follow_ (no behavior change)
3. New build steps will _not follow_ (behavior change)

I'm OK with (3) and it's something we can easily change if people don't like the inconsistency.

**Comment by jeffret-b** [comment]
Let's try kicking the CI build.

**Comment by jeffret-b** [comment]
CI passes now. Marking ready-for-merge. This can be merged after the waiting period.

**Comment by jeffret-b** [comment]
Thanks for the contribution @mwinter69!

File: bom/pom.xml
Comment (daniel-beck) [inline bom/pom.xml:62]: This change breaks `mvn hpi:run` in plugins depending on Jenkins 2.230+ with JDK 8 with the error:

> `Error resolving project artifact: Failure to transfer com.sun:tools:pom:1.8.0` [...]

Comment (jeffret-b) [reply bom/pom.xml:62]: We can probably exclude that dependency. It will take more research. Have you filed a Jira ticket or anything @daniel-beck ?

Comment (mwinter69) [reply bom/pom.xml:62]: actually ant doesn't declare this dependency at all. Could this come from the fact that this version of ant was compiled with java11 though with the class versions for java8? If yes we might go back to ant 1.10.5. The reason for using newer ant was that since version 1.10.2 (might also be 1.10.3) ant is able to properly handle symlinks on windows.

Comment (jeffret-b) [reply bom/pom.xml:62]: 1.10.5 looks like it works from my quick tests. That's a reasonable explanation for where the extra dependency comes from.

Comment (daniel-beck) [reply bom/pom.xml:62]: Filed [JENKINS-61937](https://issues.jenkins-ci.org/browse/JENKINS-61937) just now for this.

Weirdly it looked like the dependency exists when I `dependency:tree` the plugin with the Jenkins dependency.

**Pull request merged**
